### PR TITLE
Revert Serializing Outbound Transport Messages on IO Threads

### PR DESCRIPTION
**Pull request by original-brownbear**
Serializing outbound transport message on the IO loop was introduced in https://github.com/elastic/elasticsearch/pull/56961. Unfortunately it turns out that this is incompatible with assumptions made by CCR code here: https://github.com/elastic/elasticsearch/blob/f22ddf822e24bb17f1772c3bacb7ee97a00339b8/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/repositories/GetCcrRestoreFileChunkAction.java#L60-L61 and that are not easy to work around on short notice.

Raising reverting this move (as a temporary solution, it's still a valuable change long-term) as a blocker therefore as this seriously affects the stability of the initial phase of the CCR following by causing corrupted bytes to be send to the follower.

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 1]
```diff
@@ -93,15 +91,14 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
 
     @Override
     public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
-        assert msg instanceof OutboundHandler.SendContext;
-        assert Transports.assertDefaultThreadContext(transport.getThreadPool().getThreadContext());
```
Comment (DaveCTurner) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java:97]: Think we can keep this assertion? It relates to #57792.

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java [hunk 2]
```diff
@@ -115,17 +120,16 @@ private void sendMessage(TcpChannel channel, OutboundMessage networkMessage, Act
         internalSend(channel, sendContext);
     }
 
-    private void internalSend(TcpChannel channel, SendContext sendContext) {
+    private void internalSend(TcpChannel channel, SendContext sendContext) throws IOException {
         channel.getChannelStats().markAccessed(threadPool.relativeTimeInMillis());
-        // stash thread context so that channel event loop is not polluted by thread context
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/OutboundHandler.java:120]: Think we can keep this comment? It relates to #57792.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I did the revert myself fairly naively and found just a couple of small differences, see below.

**Comment by original-brownbear** [comment]
Thanks for spotting those two, sorry for the oversight!

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**